python/ray/data/datasource/datasource.py

```diff
@@ -225,7 +225,21 @@ def apply_predicate(
 class Datasource(_DatasourceProjectionPushdownMixin, _DatasourcePredicatePushdownMixin):
     """Interface for defining a custom :class:`~ray.data.Dataset` datasource.
 
+    User may subclass this class to implement a custom datasource. The subclass should
+    implement :meth:`.get_read_tasks` and
+    :meth:`.estimate_inmemory_data_size` to read the data and estimate the in-memory data size, respectively.
+
     To read a datasource into a dataset, use :meth:`~ray.data.read_datasource`.
+
+    Example:
+        >>> class MyDatasource(Datasource): # doctest: +SKIP
+        ...     def get_read_tasks(self, parallelism: int, **read_args) -> List["ReadTask"]: # doctest: +SKIP
+        ...         return [ReadTask(lambda: [pa.Table.from_pydict({"data": range(10)})], BlockMetadata(10, 80, None, None))] # doctest: +SKIP
+        ...     def estimate_inmemory_data_size(self) -> Optional[int]: # doctest: +SKIP
+        ...         return 80
+        >>> ds = MyDatasource() # doctest: +SKIP
+        >>> tasks = ds.get_read_tasks(parallelism=1) # doctest: +SKIP
+        >>> assert len(tasks) == 1 and tasks[0].metadata.num_rows == 10 # doctest: +SKIP
```

Comment:
@wingkitlee0 any reason we can't test this? If not, could we remove the `doctest: +SKIP`s?

Reply:
Also, this might be easier to read as a `.. testcode::` rather than a REPL-style example